--- /opt/mozilla/zookeeper/3.4.4/src/contrib/zkpython/src/c/zookeeper.c	2012-09-17 10:34:37.000000000 +0200
+++ zookeeper.c	2012-09-25 12:32:05.000000000 +0200
@@ -588,7 +589,8 @@
   //  int clientid = -1;
   clientid_t cid;
   cid.client_id = -1;
-  const char *passwd;
+  const char *passwd = NULL;
+  unsigned int passwd_len = 0;
   int handle = next_zhandle();
   if (handle == -1) {
     if (resize_zhandles() == 0) {
@@ -602,11 +604,15 @@
     return NULL;
   }
 
-  if (!PyArg_ParseTuple(args, "s|Oi(Ls)", &host, &watcherfn, &recv_timeout, &cid.client_id, &passwd)) 
+  if (!PyArg_ParseTuple(args, "s|Oi(Ls#)", &host, &watcherfn, &recv_timeout, &cid.client_id, &passwd, &passwd_len))
     return NULL;
-  
+
   if (cid.client_id != -1) {
-    strncpy(cid.passwd, passwd, 16*sizeof(char));
+    if (passwd_len != 16) {
+      PyErr_SetString(BadArgumentsException, "Session password should be 16 bytes long");
+      return NULL;
+    }
+    memcpy(cid.passwd, passwd, 16*sizeof(char));
   }
   pywatcher_t *pyw = NULL;
   if (watcherfn != Py_None) {
@@ -1291,7 +1297,7 @@
   }
   CHECK_ZHANDLE(zkhid);
   const clientid_t *cid = zoo_client_id(zhandles[zkhid]);
-  return Py_BuildValue("(L,s)", cid->client_id, cid->passwd);
+  return Py_BuildValue("(L,s#)", cid->client_id, cid->passwd, 16);
 }
 
 /* DO NOT USE - context is used internally. This method is not exposed
